### PR TITLE
MGMT-8348: Change pkg/upgrade/ into a typed interface

### PR DESCRIPTION
**Pull request by pmtk**
The PR changes structure of the `pkg/update/` package to expose an mockable interface and export functions used outside of that package. Newly create `ClusterInfo` interface is added as a field to the `SpecialResourceReconciler` and created in `main.go`.
Unit test for `nodeVersionInfo()` was added. `driverToolkitVersion()` should be tested after refactoring `resource` package and being able to mock it.

### CODE REVIEW

File: pkg/upgrade/upgrade.go [hunk 1]
```diff
@@ -27,9 +26,24 @@ type NodeVersion struct {
 	DriverToolkit  registry.DriverToolkitEntry `json:"driverToolkit"`
 }
 
-func ClusterInfo() (map[string]NodeVersion, error) {
+type ClusterInfo interface {
+	GetClusterInfo() (map[string]NodeVersion, error)
+}
+
+func NewClusterInfo() *clusterInfo {
```
Comment (yevgeny-shnaidman) [inline pkg/upgrade/upgrade.go]: I think it is better to return the intefrace ClusterInfo, then the inner struct clusterInfo

Comment (pmtk) [reply pkg/upgrade/upgrade.go]: You are absolutely right. This doesn't make much sense right now

File: pkg/upgrade/upgrade_test.go [hunk 2]
```diff
@@ -1,17 +1,24 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-var _ = ginkgo.Describe("pkg/upgrade", func() {
-	ginkgo.Context("UpdateInfo", func() {
+var _ = Describe("ClusterInfo", func() {
+
+	clusterInfo := NewClusterInfo().(*clusterInfo)
+
+	Context("updateInfo", func() {
```
Comment (qbarrand) [inline pkg/upgrade/upgrade_test.go]: I don't think we want to test private functions. Could we refactor the tests to integrate those cases into the public method's?

Comment (pmtk) [reply pkg/upgrade/upgrade_test.go]: That's a good idea.
The thing is that `driverToolkitVersion()` depends very heavily on package `registry` (which e.g. creates pull secret on disk, uses `github.com/google/go-containerregistry/pkg/v1` and `github.com/google/go-containerregistry/pkg/crane`).

What do you think about refactoring `registry` first and using its mock here?

Comment (qbarrand) [reply pkg/upgrade/upgrade_test.go]: That makes sense. Do you want to wait until then before we merge this?

Comment (pmtk) [reply pkg/upgrade/upgrade_test.go]: Yes. I will refactor the `registry` without UTs and after its merge use it here.

**Comment by pmtk** [comment]
I'll update this PR with new tests after merge of https://github.com/openshift-psap/special-resource-operator/pull/174 and https://github.com/openshift-psap/special-resource-operator/pull/167

File: pkg/upgrade/upgrade_test.go [hunk 3]
```diff
@@ -1,152 +1,251 @@
...
-			table.DescribeTable("test table", func(input map[string]upgrade.NodeVersion, expected string) {
-				_, err := upgrade.UpdateInfo(input, defaultDTKInput, defaultImageURLInput)
-				gomega.Expect(err).To(gomega.HaveOccurred())
-				gomega.Expect(err.Error()).To(gomega.Equal(expected))
-			}, testCases...)
+		}, testCases...)
+	})
+
+	When("NFD is not installed", func() {
+		It("will hint that with an error message", func() {
```
Comment (qbarrand) [inline pkg/upgrade/upgrade_test.go]: How about removing the `When` here?
```suggestion
	It("will hint that with an error message when NFD is not installed", func() {
```

Comment (pmtk) [reply pkg/upgrade/upgrade_test.go]: Done

**Pull request merged**